pallets/cheqd-migration/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+    #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+    #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+    #[derive(scale_info_derive::TypeInfo)]
+    #[scale_info(omit_prefix)]
+    pub struct CheqdAddress(String);
+
+    impl CheqdAddress {
+        pub fn new<T>(value: String) -> Result<Self, Error<T>> {
+            let (prefix, addr) =
+                bech32::decode(&value).map_err(|_| Error::<T>::AddressMustBeValidBech32)?;
+            ensure!(
+                prefix.as_str() == "cheqd",
+                Error::<T>::AddressMustStartWithCheqd
+            );
+            ensure!(addr.len() == 20, Error::<T>::InvalidAddressLength);
+
+            Ok(Self(value))
+        }
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        /// The currency to be burnt during migration.
+        type Currency: Currency<Self::AccountId>;
+        /// Id of the recipient account on the dock side to send burnt funds.
+        type BurnDestination: Get<Self::AccountId>;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// `cheqd` address must start with `cheqd` prefix.
+        AddressMustStartWithCheqd,
+        /// Provided `cheqd` address is invalid because it has an incorrect length.
+        InvalidAddressLength,
+        /// `cheqd` address part coming after `cheqd` must be a valid bech-32 sequence.
+        AddressMustBeValidBech32,
+        /// Caller account's balance is zero.
+        BalanceIsZero,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Burns the free `DOCK` balance of the sender and emits an event containing the supplied recipient `cheqd` address.
+        /// By submitting this transaction, you agree to the Terms and Conditions.
+        #[pallet::weight(SubstrateWeight::<T::DbWeight>::migrate())]
+        pub fn migrate(origin: OriginFor<T>, cheqd_address: String) -> DispatchResultWithPostInfo {
+            let dock_account = ensure_signed(origin)?;
+
+            let cheqd_account = CheqdAddress::new::<T>(cheqd_address)
+                .map_err(DispatchError::from)
+                .map_err(|error| {
+                    error.with_weight(SubstrateWeight::<T::DbWeight>::migrate_validation_failure())
+                })?;
+
+            let dock_tokens_amount = T::Currency::free_balance(&dock_account);
+            ensure!(!dock_tokens_amount.is_zero(), Error::<T>::BalanceIsZero);
+
+            let dest = T::BurnDestination::get();
+            T::Currency::transfer(&dock_account, &dest, dock_tokens_amount, AllowDeath)?;
+
+            Self::deposit_event(Event::Migrated {
+                dock_account,
+                cheqd_account,
+                dock_tokens_amount,
+                accepted_terms_and_conditions: true,
```

Comment:
When would this be false?

Reply:
Never; this fact should just be mentioned at the event.
Alternatively, we could have this flag as a transaction parameter, and only emit an event in case it was set to `true`.

Reply:
Whats the significance then? And if its a transaction parameter, how would the transaction behave on different values of it? Isn't calling migrate implying that caller accepts the terms?

Reply:
I was asked to add this field by @mike-parkhill. I guess that it has a demonstrative meaning.

Reply:
Ok. I still don't understand what its demonstrating.

Reply:
I see you're argument that we may not need it. But, it makes the acceptance explicit rather than implicit which is generally easier to defend if we ever get called on it. 

Going through the migration page we're building will make it mandatory as part of the flow, but for blockchain clients who want to submit it themselves we still need to get them to show they accept the terms and conditions somehow. This was my suggestion for how to do that. I'm open to other suggestions that achieve the same thing.

Reply:
With blockchain, the acceptance is always implicit, no? Eg. when you make a transfer txn, its understood that you can't ask the chain to revert that transfer. Secondly there is no way to do that txn while not accepting the terms and its not known what we would do differently if we allow that.

Anyway, I don't think its a major problem so if it makes you feel safer or more compliant, we can go ahead with this.